tests/unit/samplegen/test_template.py

```diff
@@ -0,0 +1,318 @@
+# Copyright (C) 2019  Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import jinja2
+import os.path as path
+import unittest
+import gapic.samplegen.samplegen as samplegen
+import gapic.samplegen.utils as utils
+
+
+class TestTemplating(unittest.TestCase):
+
+    def check_template(self, template_fragment, expected_output):
+        # Making a new environment for every unit test seems wasteful,
+        # but the obvious alternative (make env an instance attribute
+        # and passing a FunctionLoader whose load function returns
+        # a constantly reassigned string attribute) isn't any faster
+        # and is less clear.
+        env = jinja2.Environment(
+            loader=jinja2.ChoiceLoader(
+                [jinja2.FileSystemLoader(
+                    searchpath=path.realpath(path.join(path.dirname(__file__),
+                                                       "..", "..", "..",
+                                                       "gapic", "templates", "examples"))),
+                 jinja2.DictLoader({"template_fragment": template_fragment}),
+                 ]),
+
+            undefined=jinja2.StrictUndefined,
+            extensions=["jinja2.ext.do"],
+            trim_blocks=True, lstrip_blocks=True
+        )
+
+        env.filters["split_downcase"] = utils.split_caps_and_downcase
+
+        template = env.get_template("template_fragment")
+        text = template.render()
+        self.assertEqual(text, expected_output)
+
+    def test_handle_attr_value(self):
+        self.check_template(
+            '''{% import "feature_fragments.j2" as frags %}
```

Comment:
The `textwrap` module has a dedent function which you can use to make this nicer.

Reply:
That is so frustratingly useful and would have saved me something like three person-months of pain.